source/common/http/alternate_protocols_cache.h

```diff
@@ -73,34 +77,55 @@ class AlternateProtocolsCache {
     uint32_t port_;
   };
 
-  explicit AlternateProtocolsCache(TimeSource& time_source);
+  virtual ~AlternateProtocolsCache() = default;
 
   // Sets the possible alternative protocols which can be used to connect to the
   // specified origin. Expires after the specified expiration time.
-  void setAlternatives(const Origin& origin, const std::vector<AlternateProtocol>& protocols,
-                       const MonotonicTime& expiration);
+  virtual void setAlternatives(const Origin& origin,
+                               const std::vector<AlternateProtocol>& protocols,
+                               const MonotonicTime& expiration) PURE;
 
   // Returns the possible alternative protocols which can be used to connect to the
   // specified origin, or nullptr if not alternatives are found. The returned pointer
-  // is owned by the AlternateProtocolsCache and is valid until the next operation on
-  // AlternateProtocolsCache.
-  OptRef<const std::vector<AlternateProtocol>> findAlternatives(const Origin& origin);
+  // is owned by the AlternateProtocols and is valid until the next operation on
+  // AlternateProtocols.
+  virtual OptRef<const std::vector<AlternateProtocol>> findAlternatives(const Origin& origin) PURE;
```

Comment:
Random thoughts that I should have had on the previous CL: why an optional instead of an empty array?

Reply:
Heh. I just want to return a possibly-null pointer! *sob* :) 

So if this method didn't return an optional, we probably wouldn't want it to return the vector (as a copy) so we'd want to return reference. In the case of an empty vector, we'd need an empty vector somewhere in order to take a reference to it. That could be done, of course, but I'm not sure it's any better?

Reply:
Yeah you'd want a static const empty vector for that. But up to you, both options work.